### DISCUSSION TITLE
How do I serve files with edge functions?

### DISCUSSION THREAD

**Discussion by martin-dimi** [vercel/vercel#10160] [Help]
### Problem Description

Hi I've got a static js file in the dist folder. Is it possible to serve it with an edge function? If yes, how would I do that?

**Comment by leerob** [comment]
Is there a reason you want to use a function versus serving the file on our Edge Network statically?

**Reply by martin-dimi** [reply]
Yes, we do validation against a blacklist to whom we serve it

**Comment by louisheath** [comment]
Is there any possible solution here 🙏 We'd like to be able to serve a JS file quickly to many regions while also first validating who can download it , which is why we are using a function

TIA !

**Comment by louisheath** [comment]
Hi @leerob , we managed to move this implementation to an edge function as suggested, but with mixed success ...

- In the vite `writeBundle` hook we copy the built bundle into a js file, assigning it to a const. This file lives alongside the function `api/index.ts`
- `api/index.ts` then imports the const from the neighbouring file and serves it upon request.

This works all of the time locally, however in vercel it works only some of the time.
Sometimes the edge function build fails with error `api/index.ts(1,27): error TS2307: Cannot find module './_sdk.js' or its corresponding type declarations.`

At first I thought this was because the file has not yet been generated - that the vite hooks are asynchronous and there is a race, however I log when the file is generated and it happens well before the error.

```
17:15:52.761 | sdk:build: Writing SDK to file
17:15:52.761 | sdk:build: SDK written to file
17:15:52.761 | sdk:build: computing gzip size...
17:15:52.761 | sdk:build: dist/src/index.html               1.78 kB
17:15:52.761 | sdk:build: dist/assets/index-3e9bc34b.css   14.23 kB │ gzip:  3.81 kB
17:15:52.761 | sdk:build: dist/assets/api-87807c4b.js       3.46 kB │ gzip:  1.70 kB
17:15:52.761 | sdk:build: dist/sdk.bundle.js               14.44 kB │ gzip:  5.99 kB
17:15:52.762 | sdk:build: dist/renderer.bundle.js         201.85 kB │ gzip: 67.57 kB
17:15:52.762 | sdk:build: ✓ built in 1.80s
17:15:52.762 |  
17:15:52.762 | Tasks:    2 successful, 2 total
17:15:52.762 | Cached:    2 cached, 2 total
17:15:52.762 | Time:    872ms >>> FULL TURBO
17:15:52.762 | Summary:    /vercel/path0/.turbo/runs/2VItJcTCq4ICjgPknPgScMIYars.json
17:15:52.762 |  
17:15:52.843 | Installing dependencies...
17:15:55.760 |  
17:15:55.761 | up to date in 3s
17:15:55.761 |  
17:15:55.761 | 182 packages are looking for funding
17:15:55.761 | run `npm fund` for details
17:15:56.050 | Using TypeScript 4.9.3 (local user-provided)
17:15:58.329 | api/index.ts(1,27): error TS2307: Cannot find module './_sdk.js' or its corresponding type declarations.
17:15:58.330 |  
17:15:58.368 | Build Completed in /vercel/output [10s]
17:15:58.508 | Deploying outputs...
17:15:58.629 | Injecting preview comments middleware, you can disable this in your project settings.
17:15:58.710 | Error: The Edge Function "api/index" is referencing unsupported modules:
17:15:58.710 | - sdk/api/index.js: ./_sdk.js
```

I don't have understanding of how vercel is ordering these processes under the hood.

Are you able to share any insights that would enable me to fix the stability of this build ?

TIA 🙏
